File: frontend/src/api/endpoints/totp.ts
import type { FlaskBaseResponse } from "@/types/flask.d"
import { HttpClient } from "../httpClient"

export interface TOTPSetupResponse {
	secret: string
	otpauth_url: string
	qr_data_uri: string
	backup_codes: string[]
}

export interface TOTPStatusResponse {
	enabled: boolean
}

export interface TOTPBackupCodesResponse {
	backup_codes: string[]
}

export interface TOTPValidateResponse {
	access_token: string
	token_type: string
}

export interface TOTPValidateRequest {
	temp_token: string
	code?: string
	backup_code?: string
}

export interface TOTPDeleteRequest {
	code?: string
	backup_code?: string
}

export default {
	/** Get 2FA status for current user */
	getStatus() {
		return HttpClient.get<FlaskBaseResponse & TOTPStatusResponse>("/auth/2fa/status")
	},

	/** Start 2FA setup — get QR code and backup codes */
	setup() {
		return HttpClient.post<FlaskBaseResponse & TOTPSetupResponse>("/auth/2fa/setup")
	},

	/** Verify setup with a TOTP code to activate 2FA */
	verifySetup(code: string) {
		return HttpClient.post<FlaskBaseResponse>("/auth/2fa/verify-setup", { code })
	},

	/** Disable 2FA (requires TOTP code or backup code) */
	disable(payload: TOTPDeleteRequest) {
		return HttpClient.delete<FlaskBaseResponse>("/auth/2fa/disable", { data: payload })
	},

	/** Validate 2FA code during login (uses temp_token, no auth header) */
	validate(payload: TOTPValidateRequest) {
		return HttpClient.post<FlaskBaseResponse & TOTPValidateResponse>("/auth/2fa/validate", payload)
	},

	/** Regenerate backup codes (requires TOTP code) */
	regenerateBackupCodes(code: string) {
		return HttpClient.post<FlaskBaseResponse & TOTPBackupCodesResponse>("/auth/2fa/backup-codes/regenerate", {
			code
		})
	}
}

File: backend/app/auth/models/totp.py
"""TOTP two-factor authentication models and schemas."""

import datetime
from typing import Optional

from pydantic import BaseModel
from sqlmodel import JSON
from sqlmodel import Column
from sqlmodel import Field
from sqlmodel import SQLModel
from sqlmodel import Text


class UserTOTP(SQLModel, table=True):
    """Stores per-user TOTP 2FA configuration."""

    __tablename__ = "user_totp"

    id: Optional[int] = Field(default=None, primary_key=True)
    user_id: int = Field(unique=True, index=True)

    # Fernet-encrypted TOTP secret
    secret_enc: str = Field(sa_column=Column(Text, nullable=False))

    # Not enabled until user verifies with a valid code
    enabled: bool = Field(default=False)

    # List of {hash: str, used: bool} — bcrypt hashed backup codes
    backup_codes: list = Field(default=[], sa_column=Column(JSON, nullable=False))

    # Last TOTP counter used — prevents replay of same code
    last_used_at: Optional[int] = Field(default=None)

    created_at: datetime.datetime = Field(default_factory=datetime.datetime.utcnow)


# ── Pydantic request / response schemas ──────────────────────────────────────


class TOTPSetupResponse(BaseModel):
    """Returned on setup — contains secret, QR data, and one-time backup codes."""

    secret: str
    otpauth_url: str
    qr_data_uri: str
    backup_codes: list[str]
    message: str = "Scan the QR code with your authenticator app, then verify with a code."


class TOTPVerifyRequest(BaseModel):
    code: str


class TOTPDisableRequest(BaseModel):
    code: Optional[str] = None
    backup_code: Optional[str] = None


class TOTPValidateRequest(BaseModel):
    """Used during login — temp_token + TOTP code or backup code."""

    temp_token: str
    code: Optional[str] = None
    backup_code: Optional[str] = None


class TOTPStatusResponse(BaseModel):
    enabled: bool
    message: str = "2FA status retrieved"
    success: bool = True


class TOTPBackupCodesResponse(BaseModel):
    backup_codes: list[str]
    message: str = "New backup codes generated. Save them — they are shown only once."
    success: bool = True
